### ISSUE TITLE
PKP endpoint applies IP-based geoblocking

### ISSUE THREAD

**Issue by derhuerst** [public-transport/hafas-client#184]
> ```
> FetchError: request to https://mobil.rozklad-pkp.pl:8019/bin/mgate.exe? failed, reason: connect ETIMEDOUT 78.133.168.217:8019
> ```

Does anyone know if the endpoint just moved, or if it has been shut off? @juliuste @marudor

**Comment by vkrause** [comment]
Still seems to work here.

**Comment by derhuerst** [comment]
"here" as in your internet connection, or kpublictransport's implementation?

**Comment by vkrause** [comment]
Using KPublicTransport's implementation, using the same endpoint and resolved to the same IP.

**Comment by derhuerst** [comment]
Does this work for you?

```shell
curl -H 'accept-encoding: gzip, br, deflate' -H 'accept: application/json' -H 'content-type: application/json' -d '{"lang":"en","svcReqL":[{"cfg":{"polyEnc":"GPA"},"meth":"LocMatch","req":{"input":{"loc":{"type":"ALL","name":"wroclaw glowny?"},"maxLoc":5,"field":"S"}}}],"client":{"type":"AND","id":"HAFAS"},"ver":"1.18","auth":{"type":"AID","aid":"DrxJYtYZQpEBCtcb"}}' 'https://mobil.rozklad-pkp.pl:8019/bin/mgate.exe' -v
```

**Comment by vkrause** [comment]
Yep, returns the following (after decompression):
`{"ver":"1.18","lang":"eng","id":"mv44qr7ckmq8gk4c","err":"OK","svcResL":[{"meth":"LocMatch","err":"OK","res":{"common":{"locL":[],"prodL":[],"polyL":[],"layerL":[{"id":"standard","name":"standard","index":0,"annoCnt":0}],"crdSysL":[{"id":"standard","index":0,"type":"WGS84"}],"opL":[],"remL":[{"type":"A","code":"Pa","prio":999,"icoX":1,"txtN":"Samochód i akcesoria"}],"icoL":[{"res":"prod_ic","fg":{"r":255,"g":255,"b":255},"bg":{"r":77,"g":170,"b":221}},{"res":"INFO"},{"res":"poi_a"},{"res":"prod_ir","fg":{"r":255,"g":255,"b":255},"bg":{"r":40,"g":179,"b":31}}]},"match":{"field":"S","state":"L","locL":[{"lid":"A=1@O=Wrocław Główny@X=17037084@Y=51098075@U=55@L=005100069@B=1@p=1579694899@","type":"S","name":"Wrocław Główny","icoX":0,"extId":"005100069","state":"F","crd":{"x":17037084,"y":51098075,"layerX":0,"crdSysX":0},"pCls":14,"wt":16159},{"lid":"A=4@O=Wrocław, SIXT (Wroclaw Airport)@X=16880654@Y=51110004@U=104@L=980133206@B=1@p=1444815128@","type":"P","name":"Wrocław, SIXT (Wroclaw Airport)","icoX":2,"extId":"980133206","state":"F","crd":{"x":16880654,"y":51110004,"layerX":0,"crdSysX":0},"msgL":[{"type":"REM","remX":0,"tagL":["RES_LOC"]}]},{"lid":"A=1@O=Wrocław Leśnica@X=16866577@Y=51142509@U=51@L=005104133@B=1@p=1595061048@","type":"S","name":"Wrocław Leśnica","icoX":3,"extId":"005104133","state":"F","crd":{"x":16866577,"y":51142509,"layerX":0,"crdSysX":0},"pCls":12,"wt":10312},{"lid":"A=1@O=Wrocław Nadodrze@X=17032356@Y=51125888@U=51@L=005100255@B=1@p=1595061048@","type":"S","name":"Wrocław Nadodrze","icoX":3,"extId":"005100255","state":"F","crd":{"x":17032356,"y":51125888,"layerX":0,"crdSysX":0},"pCls":12,"wt":10312},{"lid":"A=1@O=Wrocław Mikołajów@X=16998296@Y=51115856@U=51@L=005104134@B=1@p=1595061048@","type":"S","name":"Wrocław Mikołajów","icoX":3,"extId":"005104134","state":"F","crd":{"x":16998296,"y":51115856,"layerX":0,"crdSysX":0},"pCls":12,"wt":10312}]}}}]}`

**Comment by derhuerst** [comment]
Huh, I can't `traceroute` to it, from 2 different DSL connections in Berlin. 🤷

**Comment by vkrause** [comment]
I don't get a proper traceroute either, but pinging and connecting seems just fine. Verified the above test from a second connection now too, also works fine.

**Comment by marudor** [comment]
PKP hafas in marudor.de is also broken. Times out for every request. 
(Hetzner cloud Server)

**Comment by derhuerst** [comment]
It works again from my home and from various servers. What about you @marudor?

**Comment by takenek** [comment]
@derhuerst mobil.rozklad-pkp.pl is a comercial endpoint same like rozklad-pkp.pl is a comercial webpage etc.. look there is ads and other - this is not PUBLIC webpage information!
Official public Polish railroad information webpage is: https://www.portalpasazera.pl

Best Regards
TaKeN

**Comment by derhuerst** [comment]
> mobil.rozklad-pkp.pl is a comercial endpoint same like rozklad-pkp.pl is a comercial webpage etc.. look there is ads and other - this is not PUBLIC webpage information! Official public Polish railroad information webpage is: https://www.portalpasazera.pl

Would be great to have this note in `p/pkp/readme.md`, so that `hafas-client` users know what they're dealing with. Would you mind submitting a PR @takenek?

**Comment by volo-droid** [comment]
> @derhuerst mobil.rozklad-pkp.pl is a comercial endpoint same like rozklad-pkp.pl is a comercial webpage etc.
> look there is ads and other - this is not PUBLIC webpage information!

Yes, but that's not the full story. The site is operated by [PKP Informatyka](https://en.wikipedia.org/wiki/PKP_Informatyka) which is part of PKP Group (and which has been developing tickets sale system for PKP Intercity at intercity.pl and their own at bilkom.pl). The company is privately owned AFAIU. The site isn't opening for me at the moment, but an [archived version](https://web.archive.org/web/20230412204413/https://rozklad-pkp.pl/pl/page/regulamin) of their ToS page contains the following copyright:

> © 2023 PKP Informatyka & PKP S.A.

and the following info:

> 1. Portal rozklad-pkp.pl jest zarządzany przez firmę Polskie Koleje Państwowe S.A. zwaną dalej PKP SA. System planowania podróży koleją dostępny na portalu jest administrowany przez PKP informatyka spółka z o.o. zwaną dalej PKP Informatyka.

Saying that, train schedules for Polish railways is provided not by PKP S.A. but by PKP Polskie Linie Kolejowe S.A. (if you're confused by all those "PKP _something_" companies, that's normal, PKP Group companies structure is a silly mess). And their official site for the train timetables is indeed https://www.portalpasazera.pl/ . It's not HAFAS-based so there's no easy way to fetch the info out of it. In 2017 there was a case where they were blocking a popular third-party app form accessing the data, asking for a 6'000 € one-time and 300 € monthly fee for an official connection 🤯 (the story ended with the app being unpublished by the author tired of fighting with a big company for an access to data).

**Comment by takenek** [comment]
Looks like You don`t understand.
https://mobil.rozklad-pkp.pl:8019/bin/mgate.exe? is a official gateway for Rozklad-PKP apps by Hacon but if You use "another" aplication you get ban that's why i please owner to remove this support from repo because this is illegal and braking law.

Best Regards
TaKeN

**Comment by volo-droid** [comment]
> because this is illegal and braking law

I'd kindly ask you to refrain from using terms like "illegal" unless you can back it up by the reference to the corresponding law paragraph. There's a huge difference in breaking app/site Terms of Service and breaking the law.

If you haven't noticed the README contains the following statement:

> Strictly speaking, permission is necessary to use this library with a HAFAS "mobile" endpoint.

In that sense Rozklad PKP endpoint is not different to any other HAFAS mobile endpoint, and there's absolutely no reason to remove it from the repo. Before using **any** of those endpoints to get the data one should obtain a permission from the corresponding data provider, otherwise an access might be restricted/banned. So there're two solutions: either to delete this repository completely, or to notify the users of the library about the need of requesting permission from the data provider (which has been done already in the README file).

**Comment by takenek** [comment]
I wonder if you would write the same if you were the owner of this API and people would abuse it?

**Comment by volo-droid** [comment]
Do you legally represent PKP Informatyka, and believe this repository infringes your copyright? If so, please name yourself here, and then the repo owners can work on solving the issue. If you are not a copyright owner (or owner's authorized representative), then you cannot demand removing a part of code on behalf of third party. 

p.s.: FYI, I'm fluent in Polish so there's no need for translation.

> PKP Informatyka zastrzega, iż zawarte w serwisie rozklad-pkp.pl oraz old.rozklad-pkp.pl informacje stanowią utwór w rozumieniu ustawy.

Nie jestem prawnikiem, ale nie ma tu żadnej wzmianki o serwisie mobil.rozklad-pkp.pl

**Comment by takenek** [comment]
Happy http://regulamin.rozklad-pkp.pl ?

**Comment by takenek** [comment]
"Nie jestem prawnikiem, ale nie ma tu żadnej wzmianki o serwisie mobil.rozklad-pkp.pl"

**Teraz już jest.**

Pozdrawiam

**Comment by DomeQdev** [comment]
I have gained **LEGAL** (without breaking any site's ToS) access to delays and disruptions of all trains in Poland.

Archive of delays: https://pkp-archive.2137.workers.dev/ (updated daily at 3:30 AM GMT+1)
Endpoint with all running trains: https://api.zbiorkom.live/public-api/trains/all
Filters:
- `?carriers=KM,SKM,IC,PR` - carriers, separated by commas  
- `&showCompletedTrips=true` - if you want to see all trips, including those already completed  
- `&goingThruOr=zielonka,warszawa-srodmiescie` - displays only trains that stop at one of the specified stations  
- `&goingThruAnd=zielonka,warszawa-srodmiescie` - displays only trains that stop at all the specified stations

Hope this helps :)

**Comment by fwam** [comment]
Hi, the solution actually is to use a specific UA – `Dalvik/2.1.0`

**Comment by derhuerst** [comment]
@fwam I'm having trouble getting it to work (from a German landline connection). Can you post a working command here, e.g. a `curl` call or short Python script, that demonstrates that it works?

**Comment by fwam** [comment]
@derhuerst hey, check matrix

**Comment by takenek** [comment]
@derhuerst I'm repeating this because it looks like you didn't see what I wrote.
mobil.rozklad-pkp.pl is a commercial endpoint, just like rozklad-pkp.pl is a commercial webpage. You can see ads and other commercial content there – this is not public railway information!

The official public Polish railroad information webpage is: https://www.portalpasazera.pl/

Since mobil.rozklad-pkp.pl is not an official public API but rather a commercial service, I kindly ask you to consider removing support for this endpoint from the GitHub project to avoid potential issues related to unauthorized usage.

Let me know if you need more details.

**Comment by fwam** [comment]
> [@derhuerst](https://github.com/derhuerst) I'm repeating this because it looks like you didn't see what I wrote. mobil.rozklad-pkp.pl is a commercial endpoint, just like rozklad-pkp.pl is a commercial webpage. You can see ads and other commercial content there – this is not public railway information!
> 
> The official public Polish railroad information webpage is: https://www.portalpasazera.pl/
> 
> Since mobil.rozklad-pkp.pl is not an official public API but rather a commercial service, I kindly ask you to consider removing support for this endpoint from the GitHub project to avoid potential issues related to unauthorized usage.
> 
> Let me know if you need more details.

rozkład jazdy

**Comment by DomeQdev** [comment]
@takenek ale to jest chore. Nie ma legalnego źródła na dostęp do rozkładów. Z portalu pasażera też nie można pobierać rozkładów według waszego głupiego regulaminu. Ile kosztuje dostęp do legalnego źródła danych?

**Comment by fwam** [comment]
@takenek ponowię pytanie wyżej, czy reprezentuje pan(i) spółkę PKP SA/PKP Informatyka?

**Comment by takenek** [comment]
> [@takenek](https://github.com/takenek) ponowię pytanie wyżej, czy reprezentuje pan(i) spółkę PKP SA/PKP Informatyka?

Ja nikogo nie reprezentuje tylko umiem czytać regulamin.



> [@takenek](https://github.com/takenek) ale to jest chore. Nie ma legalnego źródła na dostęp do rozkładów. Z portalu pasażera też nie można pobierać rozkładów według waszego głupiego regulaminu. Ile kosztuje dostęp do legalnego źródła danych?

Proszę napisać do PKP PLK pytanie ile kosztuje dostęp ja niestety tego nie wiem.

**Comment by fwam** [comment]
> Ja nikogo nie reprezentuje tylko umiem czytać regulamin a jak nie jestem pewny interpretacji regulaminu to się pytam właściciela strony co autor miał na myśli i aby poprawił regulamin aby był jasny i przejrzysty.

@takenek no, w takim spoko; dopóki nie ma C&D od PKP to wywalone xD – nie jestem legalistką

**Comment by DomeQdev** [comment]
@fwam No tylko były przypadki że PKP kogoś pozwało za nielegalny dostęp do danych, w sumie ciekawe skąd wiedzą z jakiego źródła są te dane...

**Comment by takenek** [comment]
> [@DomeQdev](https://github.com/DomeQdev) Newag też pozwał osoby z Dragon Sector, nie zmienia to faktu, że jest to SLAPP xD

Akcji z Newagiem nie ma co porównywać do celowego omijania zabezpieczeń i łamania regulaminu serwisu.

**Comment by volo-droid** [comment]
> Ja nikogo nie reprezentuje tylko umiem czytać regulamin a jak nie jestem pewny interpretacji regulaminu to się pytam właściciela strony co autor miał na myśli i aby poprawił regulamin aby był jasny i przejrzysty.

Bardzo ładnie, ale dla czego Pan nie pisze że ma relacje biznesowe z PKP informatyka?
https://heckin.how/notes/9y8lbh3ckpzbia5c

**Comment by fwam** [comment]
@takenek 
> Dziele się grzecznie informacją, że łamiecie prawo polskie i jeszcze macie do mnie pretensje?

dobra, fajnie, dzięki za info

**Comment by takenek** [comment]
> To PKP informatyka łamie polskie prawo utrudniając dostęp do informacji sektora publicznego. Regulamin strony NIGDY nie będzie powyżej polskiego prawa 😉

Poczytaj proszę kto wedle ustawy musi udostępniać dane rozkładu jazdy w ramach informacji publicznej w naszym kraju.

**Comment by DomeQdev** [comment]
Okej, rozkłady można sobie pobrać w formacie GTFS (którego nie udostępniają oficjalnie Koleje Mazowieckie, Polregio i Arriva), ale skąd mamy sobie wyczarować opóźnienia pociągów i numery peronów? Tego w PDFie nie ma.

**Comment by fwam** [comment]
Swoją drogą, napisanie kodu obsługującego mobil. nie jest nielegalne – nielegalne (o ile i to) jest używanie go.

**Comment by DomeQdev** [comment]
@takenek całą tą konwersację mam na mailu, nie wiem po co usunąłeś te wpisy

**Comment by fajfer** [comment]
> [@takenek](https://github.com/takenek)
> 
> Dziele się grzecznie informacją, że łamiecie prawo polskie i jeszcze macie do mnie pretensje?

Prośba o podstawę prawną

**Comment by fwam** [comment]
@takenek a za to to nieironicznie dziękuję

**Comment by surfaceflinger** [comment]
Witam . Wszystcy bioracy udzial w watku i zostawiający reakcje już są namierzani.

**Comment by makindotcc** [comment]
> [@derhuerst](https://github.com/derhuerst) mobil.rozklad-pkp.pl is a comercial endpoint same like rozklad-pkp.pl is a comercial webpage etc.. look there is ads and other - this is not PUBLIC webpage information! Official public Polish railroad information webpage is: https://www.portalpasazera.pl
> 
> Best Regards TaKeN

@takenek ja nie widze zadnych reklam, a strona jest publiczna:

![Image](https://github.com/user-attachments/assets/ddc278f8-2857-41c9-af36-528d2a97e8c9)

to po co ten cyrk?

**Comment by derhuerst** [comment]
from https://github.com/public-transport/hafas-client/pull/336#issuecomment-2659678869:

> (Responding in English because I don't speak Polish.)
> 
> I think we can agree that `hafas-client` by itself is merely a tool, used to access many different HAFAS endpoints. What is considered in line with or in violation of the terms of service of a particular HAFAS endpoint depends on _who_ uses it and _what for_.
> 
> The current state of access to Polish static (and realtime!) transport data is irrelevant for this PR, just like the question if mobil.rozklad-pkp.pl's terms of service are in line with EU regulations, and just like the question what constitutes _mass_ data retrieval. These are important discussions to have, but please have them somewhere else (e.g. in [`#opentransport:matrix.org`](https://matrix.to/#/#opentransport:matrix.org), or in [`public-transport/ideas`](https://github.com/public-transport/ideas/issues)).

**Comment by derhuerst** [comment]
#336 has been merged, so I'll close this. All parties that are *legally allowed to do so* can now access the HAFAS endpoint using `hafas-client`. We cannot (and will not) do anything about the IP blocking unfortunately.

**Comment by surfaceflinger** [comment]
PKP HAFAS now needs () after the UA, eg `Dalvik/2.1.0 ()`

I think it would be a good idea to include some real device in parenthesis, so that banning using UAs would also ban real devices. Even better if we could pick a random UA from a list of devices, so they have to switch to a different method.

@derhuerst could you reopen?

**Comment by fwam** [comment]
przypominam, że pan Marek to czyta xD

**Comment by surfaceflinger** [comment]
It changed again today, empty () doesn't work anymore so as I said, ideally there should be a list of user agents from the most popular Android phones so that blocking by UAs becomes unreliable.

Not sure if it's possible for them to track if subsequent requests belong to the same device (by matching IP?) so UA should stay the same on each startup for every request and switch to a different one once there are multiple failures.

**Comment by fwam** [comment]
@takenek co ci tak zależy? serio xD
